flusher.go

```diff
@@ -370,3 +377,108 @@ func (s *Server) flushForward(ctx context.Context, wms []WorkerMetrics) {
 func (s *Server) flushTraces(ctx context.Context) {
 	s.SpanWorker.Flush()
 }
+
+// forwardGRPC forwards all input metrics to a downstream Veneur, over gRPC.
+func (s *Server) forwardGRPC(ctx context.Context, wms []WorkerMetrics) {
+	span, _ := trace.StartSpanFromContext(ctx, "")
+	span.SetTag("protocol", "grpc")
+	defer span.ClientFinish(s.TraceClient)
+
+	exportStart := time.Now()
+	metrics := s.exportForwardMetrics(wms)
+	span.Add(
+		ssf.Timing("forward.duration_ns", time.Since(exportStart),
+			time.Nanosecond, map[string]string{"part": "export"}),
+		ssf.Gauge("forward.post_metrics_total", float32(len(metrics)), nil),
+	)
+
+	if len(metrics) == 0 {
+		log.Debug("Nothing to forward, skipping.")
+		return
+	}
+
+	entry := log.WithFields(logrus.Fields{
+		"metrics":     len(metrics),
+		"destination": s.ForwardAddr,
+		"protocol":    "grpc",
+		"grpcstate":   s.grpcForwardConn.GetState().String(),
+	})
+
+	c := forwardrpc.NewForwardClient(s.grpcForwardConn)
+
+	grpcStart := time.Now()
+	_, err := c.SendMetrics(ctx, &forwardrpc.MetricList{Metrics: metrics})
+	if err != nil {
+		span.Add(ssf.Count("forward.error_total", 1, map[string]string{"cause": "send"}))
+		entry.WithError(err).Error("Failed to forward to an upstream Veneur")
+	} else {
+		entry.Info("Completed forward to an upstream Veneur")
+	}
+
+	span.Add(
+		ssf.Timing("forward.duration_ns", time.Since(grpcStart), time.Nanosecond,
+			map[string]string{"part": "grpc"}),
+		ssf.Count("forward.error_total", 0, nil),
+	)
+}
+
+// exportForwardedMetrics converts a slice of WorkerMetrics to a slice
+// of protobuf-compatible metrics for use in forwarding over gRPC
+func (s *Server) exportForwardMetrics(wms []WorkerMetrics) []*metricpb.Metric {
+	bufLen := 0
+	for _, wm := range wms {
+		bufLen += len(wm.histograms) + len(wm.sets) + len(wm.timers) +
+			len(wm.globalCounters) + len(wm.globalGauges)
+	}
+
+	metrics := make([]*metricpb.Metric, 0, bufLen)
```

Comment:
I suppose that could fit here, but it seems like it might be a bit of an unnecessary optimization?  This method is called once each flush, which happens (usually) every 10 seconds and there should only be one flush executing at once.  At that low of a frequency it also seems likely that these slices would get GC'ed between each flush anyway?

Reply:
ahh - i missed that it was only called via flush vectors! yes, that alleviates my concern - but let's put a comment on it noting that if this is ever to be called from anything that's on a hot path, it should be poolified.